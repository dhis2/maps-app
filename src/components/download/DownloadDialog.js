import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import i18n from '@dhis2/d2-i18n';
import {
    Modal,
    ModalTitle,
    ModalContent,
    ModalActions,
    Button,
    ButtonStrip,
} from '@dhis2/ui';
import Checkbox from '../core/Checkbox';
import LegendPosition from './LegendPosition';
import {
    toggleDownloadDialog,
    toggleDownloadShowName,
    toggleDownloadShowLegend,
    setDownloadLegendPosition,
} from '../../actions/download';
import {
    convertToPng,
    dataURItoBlob,
    downloadFile,
    downloadSupport,
} from '../../util/export-image';
import styles from './styles/DownloadDialog.module.css';

export class DownloadDialog extends Component {
    static propTypes = {
        showDialog: PropTypes.bool.isRequired,
        showName: PropTypes.bool.isRequired,
        showLegend: PropTypes.bool.isRequired,
        legendPosition: PropTypes.string.isRequired,
        hasName: PropTypes.bool.isRequired,
        hasLegend: PropTypes.bool.isRequired,
        toggleDownloadDialog: PropTypes.func.isRequired,
        toggleDownloadShowName: PropTypes.func.isRequired,
        toggleDownloadShowLegend: PropTypes.func.isRequired,
        setDownloadLegendPosition: PropTypes.func.isRequired,
    };

    state = {
        error: null,
    };

    render() {
        const {
            showDialog,
            hasName,
            showName,
            hasLegend,
            showLegend,
            legendPosition,
            toggleDownloadShowName,
            toggleDownloadShowLegend,
            setDownloadLegendPosition,
        } = this.props;

        if (!showDialog) {
            return null;
        }

        const isSupported = downloadSupport() && !this.state.error;

        return (
            <Modal position="middle" small onClose={this.onClose}>
                <ModalTitle>{i18n.t('Download map')}</ModalTitle>
                <ModalContent>
                    <div className={styles.modalContent}>
                        {isSupported ? (
                            <Fragment>
                                <Checkbox
                                    label={i18n.t('Show name')}
                                    checked={showName}
                                    disabled={!hasName}
                                    onChange={toggleDownloadShowName}
                                />
                                <Checkbox
                                    label={i18n.t('Show legend')}
                                    checked={showLegend}
                                    disabled={!hasLegend}
                                    onChange={toggleDownloadShowLegend}
                                />
                                {hasLegend && showLegend && (
                                    <LegendPosition
                                        position={legendPosition}
                                        onChange={setDownloadLegendPosition}
                                    />
                                )}
                            </Fragment>
                        ) : (
                            i18n.t(
                                'Map download is not supported by your browser. Try Google Chrome or Firefox.'
                            )
                        )}
                    </div>
                </ModalContent>
                <ModalActions>
                    <ButtonStrip end>
                        <Button secondary onClick={this.onClose}>
                            {isSupported ? i18n.t('Cancel') : i18n.t('Close')}
                        </Button>
                        {isSupported && (
                            <Button primary onClick={this.onDownload}>
                                {i18n.t('Download')}
                            </Button>
                        )}
                    </ButtonStrip>
                </ModalActions>
            </Modal>
        );
    }

    onClose = () => this.props.toggleDownloadDialog(false);

    onDownload = () => {
        const mapEl = document.getElementById('dhis2-map-container');

        const filename = `map-${Math.random()
            .toString(36)
            .substring(7)}.png`;

        // Skip map controls in download except attribution and scale
        // Mapbox map controls contains inline SVG for CSS background-image, which
        // is not accepted by dom-to-image
        // Bing Maps logo is blocked by CORS policy
        const skipElements = el =>
            !el.classList ||
            el.classList.contains('mapboxgl-ctrl-scale') ||
            el.classList.contains('mapboxgl-ctrl-attrib') ||
            !(
                el.classList.contains('mapboxgl-ctrl') ||
                el.classList.contains('dhis2-map-bing-logo')
            );

        const options = {
            width: mapEl.offsetWidth,
            height: mapEl.offsetHeight,
            filter: skipElements,
        };

        // Adding 1 to the computed width of the title element avoids text
        // wrapping outside the box in the generated image
        const titleEl = mapEl.getElementsByClassName('dhis2-maps-title')[0];

        if (titleEl) {
            const width = window
                .getComputedStyle(titleEl, null)
                .getPropertyValue('width');

            titleEl.style.width = parseInt(width, 10) + 1 + 'px';
        }

        convertToPng(mapEl, options)
            .then(dataUri => {
                downloadFile(dataURItoBlob(dataUri), filename);

                if (titleEl) {
                    titleEl.style.width = 'auto';
                }

                this.onClose();
            })
            .catch(this.onError);
    };

    // Not working in Safari: https://github.com/tsayen/dom-to-image/issues/27
    onError = error => {
        this.setState({ error });
    };
}

export default connect(
    state => ({
        hasName: state.map.name !== undefined,
        hasLegend: state.map.mapViews.filter(layer => layer.legend).length > 0,
        ...state.download,
    }),
    {
        toggleDownloadDialog,
        toggleDownloadShowName,
        toggleDownloadShowLegend,
        setDownloadLegendPosition,
    }
)(DownloadDialog);
