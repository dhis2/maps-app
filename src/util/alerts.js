// Returns array of all alerts from map config and layers (mapViews)
export const getMapAlerts = mapConfig =>
    []
        .concat(...(mapConfig && mapConfig.alerts ? mapConfig.alerts : []))
        .concat(
            ...(mapConfig &&
                mapConfig.mapViews &&
                mapConfig.mapViews
                    .filter(layer => layer.alerts)
                    .map(layer => layer.alerts))
        );
