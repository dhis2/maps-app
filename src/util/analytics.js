import i18n from '@dhis2/d2-i18n'
import { sortBy } from 'lodash/fp'
import { dimConf } from '../constants/dimension.js'
import {
    RENDERING_STRATEGY_TIMELINE,
    RENDERING_STRATEGY_SPLIT_BY_PERIOD,
} from '../constants/layers.js'
import { getPeriodNames } from './periods.js'
import { isValidUid } from './uid.js'

/* DIMENSIONS */

const createDimension = (dimension, items, props) => ({
    dimension,
    items,
    ...props,
})

const getDimension = (dimension, arr) =>
    arr.filter((item) => item.dimension === dimension)[0]

const getDimensionItems = (dimension, arr) => {
    const dataItems = getDimension(dimension, arr)
    return dataItems && dataItems.items ? dataItems.items : []
}

/* DATA ITEM */

export const getDataItemFromColumns = (columns = []) =>
    getDimensionItems('dx', columns)[0]

export const setDataItemInColumns = (dataItem, dimension) => {
    const dim = dimConf[dimension]
    return dim
        ? [
              createDimension(
                  dim.dimensionName,
                  [
                      {
                          id:
                              dataItem.id +
                              (dimension === 'reportingRate'
                                  ? '.REPORTING_RATE'
                                  : ''),
                          name: dataItem.name,
                          dimensionItemType: dim.itemType,
                          legendSet: dataItem.legendSet, // TODO: Keep outside of columns?
                      },
                  ],
                  { objectName: dim.objectName }
              ),
          ]
        : []
}

/* ORGANISATION UNIT */

export const getOrgUnitsFromRows = (rows = []) =>
    getDimensionItems('ou', rows) || []

/* ORGANISATION UNIT NODES */

const isOrgUnitNode = (ou) => isValidUid(ou.id)

export const getOrgUnitNodesFromRows = (rows = []) =>
    getOrgUnitsFromRows(rows).filter(isOrgUnitNode)

/* ORGANISATION UNIT PATH */
// Set organisation unit tree path (temporary solution, as favorites don't include paths)
export const setOrgUnitPathInRows = (rows = [], id, path) => {
    const orgUnits = getOrgUnitsFromRows(rows)
    const orgUnit = orgUnits.find((ou) => ou.id === id)
    return [
        createDimension('ou', [
            ...orgUnits.filter((ou) => ou.id !== id),
            { ...orgUnit, path },
        ]),
    ]
}

/* PERIODS */

export const getPeriodFromFilters = (filters = []) =>
    getDimensionItems('pe', filters)[0]

export const removePeriodFromFilters = (filters = []) => [
    ...filters.filter((f) => f.dimension !== 'pe'),
]

export const getPeriodNameFromId = (id) => getPeriodNames()[id]

export const setFiltersFromPeriod = (filters, period) => [
    ...removePeriodFromFilters(filters),
    createDimension('pe', [{ ...period }]),
]

/* DYNAMIC DIMENSION FILTERS */

export const getDimensionsFromFilters = (filters = []) =>
    filters.filter((d) => isValidUid(d.dimension) || d.dimension === null)

export const getValidDimensionsFromFilters = (filters = []) =>
    filters.filter((d) => isValidUid(d.dimension) && d.items && d.items.length)

export const removeDimensionFromFilters = (filters, index) => {
    const dimensions = getDimensionsFromFilters(filters)

    if (!dimensions || !dimensions[index]) {
        return filters
    }

    return [
        ...filters.filter((f) => f.dimension === 'pe'),
        ...dimensions.filter((d, i) => i !== index),
    ]
}

export const changeDimensionInFilters = (filters, index, filter) => {
    const dimensions = getDimensionsFromFilters(filters)

    if (!dimensions || !dimensions[index]) {
        return filters
    }

    dimensions[index] = filter

    return [...filters.filter((f) => f.dimension === 'pe'), ...dimensions]
}

/* FILTERS */

export const getFiltersFromColumns = (columns = []) => {
    const filters = columns.filter((item) => item.filter)
    return filters.length ? filters : null
}

// Returns list of filters in a readable format
export const getFiltersAsText = (filters = [], names = {}) =>
    filters.map(({ dimension, filter }) => {
        const [operator, value] = filter.split(':')
        const filterName = names[dimension]
        const filterOperator = getFilterOperatorAsText(operator, value)
        const filterValue = getFilterValueName(value, operator, names)

        return `${filterName} ${filterOperator} ${filterValue}`
    })

// Returns filter operator in a readable format
export const getFilterOperatorAsText = (operator, value) => {
    // If only one value, use is / is not
    if (!value.includes(';')) {
        if (operator === 'IN') {
            return i18n.t('is')
        } else if (operator === '!IN') {
            return i18n.t('is not')
        }
    }

    return {
        EQ: '=',
        GT: '>',
        GE: '>=',
        LT: '<',
        LE: '<=',
        NE: '!=',
        IN: i18n.t('one of'),
        '!IN': i18n.t('not one of'),
        LIKE: i18n.t('contains'),
        '!LIKE': i18n.t("doesn't contains"),
    }[operator]
}

// Returns filter value is a readable fromat
export const getFilterValueName = (value, operator, names) => {
    if (operator === 'IN') {
        if (value === '1') {
            return i18n.t('true')
        } else if (value === '0') {
            return i18n.t('false')
        }
    }

    return value
        .split(';')
        .map((val) => names[val] || val)
        .join(', ')
}

// Combine data items into one array and exclude certain value types
/* eslint-disable max-params */
export const combineDataItems = (
    dataItemsA = [],
    dataItemsB = [],
    includeTypes,
    excludeTypes
) =>
    sortBy(
        'name',
        [...dataItemsA, ...dataItemsB].filter(
            (item) =>
                (!includeTypes || includeTypes.includes(item.valueType)) &&
                (!excludeTypes || !excludeTypes.includes(item.valueType))
        )
    )

/* eslint-enable max-params */

// TODO: This is VERY expensive because metaData.items can have 100000+ elements.  Consider removing.
// Builds an object with key/names pairs from an API response
export const getApiResponseNames = ({ metaData, headers }) => ({
    ...Object.keys(metaData.items).reduce(
        (names, key) => ({
            ...names,
            [key]: metaData.items[key].name,
        }),
        {}
    ),
    ...headers.reduce(
        (names, header) => ({
            ...names,
            [header.name]: header.column,
        }),
        {}
    ),
    true: i18n.t('Yes'),
    false: i18n.t('No'),
})

// Returns the rendering strategy used
export const getRenderingStrategy = ({ mapViews }) => {
    if (Array.isArray(mapViews)) {
        if (
            mapViews.some(
                (view) => view.renderingStrategy === RENDERING_STRATEGY_TIMELINE
            )
        ) {
            return 'timeline'
        } else if (
            mapViews.some(
                (view) =>
                    view.renderingStrategy ===
                    RENDERING_STRATEGY_SPLIT_BY_PERIOD
            )
        ) {
            return 'split-by-period'
        }
    }
    return 'single'
}
