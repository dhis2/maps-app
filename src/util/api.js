import isString from 'lodash/fp/isString';
import isObject from 'lodash/fp/isObject';
import { config } from 'd2/lib/d2';

// TODO: Channel all api request through d2

export const apiFetch = async (url, method, body) => {
    const options = {
        headers: {},
    };

    if (config.context && config.context.auth) {
        options.headers['Authorization'] = 'Basic ' + btoa(config.context.auth);
    } else {
        options.credentials = 'include';
    }

    if (method && body) {
        options.method = method;

        if (isString(body)) {
            options.headers['Content-Type'] = 'text/html';
            options.body = body;
        } else if (isObject(body)) {
            options.headers['Content-Type'] = 'application/json';
            options.body = JSON.stringify(body);
        }
    }

    return fetch(encodeURI(config.baseUrl + url), options)
        .then(
            response =>
                method !== 'PUT' && method !== 'PATCH' && method !== 'DELETE'
                    ? response.json()
                    : response
        ) // Avoids error
        .catch(error => console.log('Error: ', error)); // TODO: Better error handling
};
