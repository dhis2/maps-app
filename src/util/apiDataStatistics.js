export const dataStatisticsMutation = {
    resource: 'dataStatistics',
    params: ({ id }) => ({
        favorite: id,
        eventType: 'MAP_VIEW',
    }),
    type: 'create',
}
