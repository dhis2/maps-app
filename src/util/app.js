import { defaultBasemaps } from '../constants/basemaps.js'
import { BING_LAYER } from '../constants/layers.js'
import {
    DEFAULT_SYSTEM_SETTINGS,
    SYSTEM_SETTINGS,
    getHiddenPeriods,
} from '../constants/settings.js'
import { createExternalLayer } from './external.js'
import { getDefaultLayerTypes } from './getDefaultLayerTypes.js'
import { fetchExternalLayersQuery } from './requests.js'

export const appQueries = {
    currentUser: {
        resource: 'me',
        params: {
            fields: 'id,username,displayName~rename(name),authorities,settings[keyAnalysisDisplayProperty]',
        },
    },
    systemSettings: {
        resource: 'systemSettings',
        params: {
            key: SYSTEM_SETTINGS,
        },
    },
    externalMapLayers: fetchExternalLayersQuery,
}

const getBasemapList = (externalMapLayers, systemSettings) => {
    const externalBasemaps = externalMapLayers
        .filter((layer) => layer.mapLayerPosition === 'BASEMAP')
        .map(createExternalLayer)

    return defaultBasemaps()
        .filter((basemap) =>
            !systemSettings.keyBingMapsApiKey
                ? basemap.config.type !== BING_LAYER
                : true
        )
        .map((basemap) => {
            if (basemap.config.type === BING_LAYER) {
                basemap.config.apiKey = systemSettings.keyBingMapsApiKey
            }
            return basemap
        })
        .concat(externalBasemaps)
}

const getLayerTypes = (externalMapLayers) => {
    const externalLayerTypes = externalMapLayers
        .filter((layer) => layer.mapLayerPosition !== 'BASEMAP')
        .map(createExternalLayer)

    return getDefaultLayerTypes().concat(externalLayerTypes)
}

export const providerDataTransformation = ({
    currentUser,
    systemSettings,
    externalMapLayers,
}) => ({
    currentUser: {
        id: currentUser.id,
        name: currentUser.name,
        username: currentUser.username,
        authorities: new Set(currentUser.authorities),
    },
    nameProperty:
        currentUser.settings.keyAnalysisDisplayProperty === 'name'
            ? 'displayName'
            : 'displayShortName',
    systemSettings: Object.assign({}, DEFAULT_SYSTEM_SETTINGS, systemSettings, {
        hiddenPeriods: getHiddenPeriods(systemSettings),
    }),
    basemaps: getBasemapList(
        externalMapLayers.externalMapLayers,
        systemSettings
    ),
    layerTypes: getLayerTypes(externalMapLayers.externalMapLayers),
})
