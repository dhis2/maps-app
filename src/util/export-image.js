import FileSaver from 'file-saver';
import { toBlob } from 'html-to-image';

export const downloadMapImage = (sourceEl, filename) =>
    toBlob(sourceEl).then(blob => FileSaver.saveAs(blob, filename));

const toStringFn = {}.toString;

// https://github.com/bubkoo/html-to-image#browsers
// https://github.com/Modernizr/Modernizr/blob/master/feature-detects/svg/foreignobject.js
// TODO: Safari is not supported, as it uses a stricter security model on <foreignObject> tag
export const downloadSupport = () => {
    return (
        !!document.createElementNS &&
        /SVGForeignObject/.test(
            toStringFn.call(
                document.createElementNS(
                    'http://www.w3.org/2000/svg',
                    'foreignObject'
                )
            )
        )
    );
};
