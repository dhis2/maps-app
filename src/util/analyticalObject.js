import { THEMATIC_LAYER } from '../constants/layers.js'
import { getPeriodNameFromId, getDimensionsFromFilters } from './analytics.js'
import { cleanDimension } from './favorites.js'
import { loadDataItemLegendSet } from './legend.js'

export const USER_DATASTORE_NAMESPACE = 'analytics'
export const CURRENT_AO_KEY = 'currentAnalyticalObject'

export const APP_URLS = {
    CHART: 'dhis-web-data-visualizer',
    PIVOT: 'dhis-web-pivot',
}

// Combines all dimensions in columns, rows and filters
const getDimensionsFromAnalyticalObject = (ao = {}) => {
    const { columns = [], rows = [], filters = [] } = ao
    return [...columns, ...rows, ...filters]
}

// Returns the data items of the first dx dimension in an analytical object
export const getDataDimensionsFromAnalyticalObject = (ao) => {
    const dims = getDimensionsFromAnalyticalObject(ao)
    const dataDim = dims.find((i) => i.dimension === 'dx')

    // We only use the first dx dimension
    return dataDim ? dataDim.items : []
}

// Returns a thematic layer config from an analytical object
export const getThematicLayerFromAnalyticalObject = async (
    ao = {},
    dataId,
    isVisible = true
) => {
    const { yearlySeries, aggregationType = 'DEFAULT' } = ao
    const dataDims = getDataDimensionsFromAnalyticalObject(ao)
    const dims = getDimensionsFromAnalyticalObject(ao)
    const orgUnits = dims.find((i) => i.dimension === 'ou')
    const filters = getDimensionsFromFilters(dims) // Dynamic dimension filters
    let period = dims.find((i) => i.dimension === 'pe')
    let dataDim = dataDims[0]

    if (dataId) {
        dataDim = dataDims.find((item) => item.id === dataId)
    }

    if (!dataDim || !orgUnits || !period) {
        return
    }

    // Load default legend set for selected data dimension
    const legendSet = await loadDataItemLegendSet(dataDim)

    // Currently we only support one period in map filters so we select the first
    if (yearlySeries && yearlySeries.length) {
        period = {
            dimension: 'pe',
            items: [
                {
                    id: yearlySeries[0],
                    name: getPeriodNameFromId(yearlySeries[0]),
                },
            ],
        }
    }

    return {
        layer: THEMATIC_LAYER,
        columns: [{ dimension: 'dx', items: [dataDim] }],
        rows: [orgUnits],
        filters: [period, ...filters],
        aggregationType,
        legendSet,
        isVisible,
        opacity: 0.9,
    }
}

// Translates a thematic layer to an analytical object
export const getAnalyticalObjectFromThematicLayer = (layer = {}) => {
    const {
        columns = [],
        rows = [],
        filters = [],
        aggregationType = 'DEFAULT',
    } = layer

    return {
        columns,
        rows: rows.map(cleanDimension),
        filters,
        aggregationType,
    }
}
