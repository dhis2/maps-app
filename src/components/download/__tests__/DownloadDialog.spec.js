import React from 'react';
import { shallow } from 'enzyme';
import { DownloadDialog } from '../DownloadDialog';
import { downloadSupport } from '../../../util/export-image';

jest.mock('../../../util/export-image', () => ({
    downloadSupport: jest.fn(),
}));

describe('DownloadDialog', () => {
    const renderComponent = (props = {}, isSupported = true) => {
        downloadSupport.mockImplementation(() => isSupported);

        return shallow(
            <DownloadDialog
                classes={{}}
                showDialog={true}
                showName={false}
                showLegend={false}
                legendPosition="bottomright"
                hasName={false}
                hasLegend={false}
                toggleDownloadDialog={() => null}
                toggleDownloadShowName={() => null}
                toggleDownloadShowLegend={() => null}
                setDownloadLegendPosition={() => null}
                {...props}
            />
        );
    };

    it('Should render null when showDialog is false', () => {
        const wrapper = renderComponent({
            showDialog: false,
        });
        expect(wrapper.isEmptyRender()).toBe(true);
    });

    it('Should render a dialog when showDialog is true', () => {
        const wrapper = renderComponent();
        expect(wrapper.isEmptyRender()).toBe(false);
        expect(wrapper.find('Modal').exists()).toBe(true);
    });

    it('Should match snapshot when showDialog is true', () => {
        const wrapper = renderComponent();
        expect(wrapper).toMatchSnapshot();
    });

    it('should call toggleDownloadDialog with false on dialog close', () => {
        const toggleDownloadDialogSpy = jest.fn();
        const wrapper = renderComponent({
            hasName: true,
            toggleDownloadDialog: toggleDownloadDialogSpy,
        });

        wrapper.simulate('close');
        expect(toggleDownloadDialogSpy).toHaveBeenCalledWith(false);
    });

    it('Should not show download options if no browser support', () => {
        const wrapper = renderComponent(null, false);

        expect(
            wrapper
                .find('ModalContent')
                .render()
                .text()
                .substring(0, 29)
        ).toEqual('Map download is not supported');

        expect(
            wrapper
                .find('Button')
                .at(0)
                .render()
                .text()
        ).toEqual('Close');
    });

    it('Should show download options if browser support', () => {
        const wrapper = renderComponent();

        expect(wrapper.find('Checkbox').length).toBe(2);
        expect(wrapper.find('Button').length).toBe(2);
        expect(
            wrapper
                .find('Button')
                .at(0)
                .render()
                .text()
        ).toEqual('Cancel');
    });

    it('Should enable name checkbox if name exist', () => {
        const wrapper = renderComponent();
        expect(
            wrapper
                .find('Checkbox')
                .at(0)
                .prop('disabled')
        ).toBe(true);
        wrapper.setProps({ hasName: true });
        expect(
            wrapper
                .find('Checkbox')
                .at(0)
                .prop('disabled')
        ).toBe(false);
    });

    it('Should check name checkbox if showName prop is true', () => {
        const wrapper = renderComponent({
            hasName: true,
        });

        expect(
            wrapper
                .find('Checkbox')
                .at(0)
                .prop('checked')
        ).toBe(false);
        wrapper.setProps({ showName: true });
        expect(
            wrapper
                .find('Checkbox')
                .at(0)
                .prop('checked')
        ).toBe(true);
    });

    it('should call toggleDownloadShowName when name checkbox is checked', () => {
        const toggleDownloadShowNameSpy = jest.fn();
        const wrapper = renderComponent({
            hasName: true,
            toggleDownloadShowName: toggleDownloadShowNameSpy,
        });
        const checkbox = wrapper.find('Checkbox').at(0);

        checkbox.simulate('change', { checked: true });
        expect(toggleDownloadShowNameSpy).toHaveBeenCalledWith(true);

        checkbox.simulate('change', { checked: false });
        expect(toggleDownloadShowNameSpy).toHaveBeenCalledWith(false);
    });

    it('Should enable legend checkbox if legend exist', () => {
        const wrapper = renderComponent();
        expect(
            wrapper
                .find('Checkbox')
                .at(1)
                .prop('disabled')
        ).toBe(true);
        wrapper.setProps({ hasLegend: true });
        expect(
            wrapper
                .find('Checkbox')
                .at(1)
                .prop('disabled')
        ).toBe(false);
    });

    it('Should check legend checkbox if showLegend prop is true', () => {
        const wrapper = renderComponent({
            hasLegend: true,
        });

        expect(
            wrapper
                .find('Checkbox')
                .at(1)
                .prop('checked')
        ).toBe(false);
        wrapper.setProps({ showLegend: true });
        expect(
            wrapper
                .find('Checkbox')
                .at(1)
                .prop('checked')
        ).toBe(true);
    });

    it('should call toggleDownloadShowLegend when name checkbox is checked', () => {
        const toggleDownloadShowLegendSpy = jest.fn();
        const wrapper = renderComponent({
            hasName: true,
            toggleDownloadShowLegend: toggleDownloadShowLegendSpy,
        });
        const checkbox = wrapper.find('Checkbox').at(1);

        checkbox.simulate('change', { checked: true });
        expect(toggleDownloadShowLegendSpy).toHaveBeenCalledWith(true);

        checkbox.simulate('change', { checked: false });
        expect(toggleDownloadShowLegendSpy).toHaveBeenCalledWith(false);
    });

    it('should show legend position if legend exist and is shown', () => {
        const wrapper = renderComponent({
            hasLegend: true,
            showLegend: false,
        });

        expect(wrapper.find('LegendPosition').exists()).toBe(false);
        wrapper.setProps({ showLegend: true });

        expect(wrapper.find('LegendPosition').exists()).toBe(true);
    });

    it('should call setDownloadLegendPosition when legend position is changed', () => {
        const setDownloadLegendPositionSpy = jest.fn();
        const wrapper = renderComponent({
            hasLegend: true,
            showLegend: true,
            setDownloadLegendPosition: setDownloadLegendPositionSpy,
        });

        wrapper.find('LegendPosition').simulate('change', 'bottomleft');
        expect(setDownloadLegendPositionSpy).toHaveBeenCalledWith('bottomleft');
    });

    it('Should show browser message if error state is set', () => {
        const wrapper = renderComponent();

        wrapper.setState({ error: {} });

        expect(
            wrapper
                .find('ModalContent')
                .render()
                .text()
                .substring(0, 29)
        ).toEqual('Map download is not supported');
    });
});
