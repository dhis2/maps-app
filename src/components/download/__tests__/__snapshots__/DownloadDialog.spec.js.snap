// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadDialog Should match snapshot when showDialog is true 1`] = `
<WithStyles(Dialog)
  onClose={[Function]}
  open={true}
>
  <WithStyles(DialogTitle)
    disableTypography={true}
  >
    Download map
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <div>
      <WithStyles(Checkbox)
        checked={false}
        disabled={true}
        label="Show name"
        onCheck={[Function]}
      />
    </div>
    <div>
      <WithStyles(Checkbox)
        checked={false}
        disabled={true}
        label="Show legend"
        onCheck={[Function]}
      />
    </div>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
      variant="contained"
    >
      Download
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;
